### PR TITLE
Add container garbage collection

### LINKED ISSUE #49: Potential for Leaked Mininet Containers
If the frontend does not make calls to terminate Mininet containers, the backend currently leaves them indefinitely. We should implement some kind of periodic garbage collection that will terminate containers associated with inactive sessions.

Furthermore, the resource checking functionality does not properly guard against creating Mininet containers when resources do not exist to do so. Profiling resource usage with many instances (observing degraded performance) will enable us to implement additional resource checks. Resource checking does need to prevent Mininet container creation when resources become constrained.

### PR DESCRIPTION
**Pull request by blazebissar**
Checkout the [Engineer Workflow Doc](https://docs.google.com/presentation/d/1pdzuHG-yO7_qW6jiByaGTLztKZ6b72lBNqSdNT5WZQ8/edit) to make sure you're contributing correctly! :)

**[Tickets](https://trello.com/b/dYOQKJGv/reclass)** 
Trello Ticket: None, closes #49 
Additional Notes: Addresses the container leak problem discussed in calls

**[Dependencies](https://docs.google.com/document/d/1CM6OA2nlhI3DHnHdDEgySukx_tZ_wsG-C2YgraZRvJA/edit#heading=h.czpk4cbi7z4g)**
None.

**[Explain Impact on DB](https://docs.google.com/document/d/1CM6OA2nlhI3DHnHdDEgySukx_tZ_wsG-C2YgraZRvJA/edit#heading=h.285lgcs4z84o)**
None.

**[Testing](https://docs.google.com/document/d/1CM6OA2nlhI3DHnHdDEgySukx_tZ_wsG-C2YgraZRvJA/edit#heading=h.rs692kr58cf8)**

How did you test this? List of actions & expected behavior with helpful screenshots for reviewer:
Manual testing creating containers and posting commands while watching `docker ps`.

If this change is performance related, please document what tests were / should be done to verify improvement:

### CODE REVIEW

File: flasksrc/session.py [hunk 1]
```diff
@@ -168,23 +173,76 @@ def getManager():
...
+        # reduce lateny if many containers expire
+        tmp_queue = State.reaper_queue.copy()
+        killed = 0
+        for container_name in tmp_queue:
+            res = State.manager.killContainer(container_name)
+            if res:
+                State.reaper_queue.remove(container_name)
+                killed += 1
+            if killed > 1:
+                break
```
Comment (srichardson40) [inline flasksrc/session.py:234]: Can we print that we killed a container successfully? That way we can verify this in the logs as well as checking docker ps

**Review by srichardson40 — APPROVED** [review]

File: flasksrc/session.py [hunk 2]
```diff
@@ -168,23 +173,76 @@ def getManager():
...
-            if res:
-                del State.containers[session]
+            if not res:
+                print(f'Failed to terminate container for session {session}')
+                # Enqueue to hopefully delete later
+                State.reaper_queue(session, None)
+
+            # Invalidate container references for that session
+            State.containers.pop(session, None)
+            State.last_used.pop(session, None)
```
[the remark this answers is not in the document]
Comment (blazebissar) [reply flasksrc/session.py]: It prevents `KeyError`s when the key doesn't exist in the dictionary, which might happen if states diverge.

File: flasksrc/__init__.py [hunk 3]
```diff
@@ -27,7 +29,9 @@ def get(self):
 
 def create_app(test_config=None):
     # create and configure the app
+    global app
     app = Flask(__name__, instance_relative_config=True)
```
[the remark this answers is not in the document]
Comment (blazebissar) [reply flasksrc/__init__.py:33]: It didn't; rather than spend a long time debugging I moved it back.

File: flasksrc/session.py [hunk 4]
```diff
@@ -168,23 +173,76 @@ def getManager():
...
+                if now - a_time > SESSION_TIMEOUT:
+                    State.reaper_queue.append(State.containers[session])
+                    State.containers.pop(session, None)
+                    State.last_used.pop(session, None)
+
+        # Iterate through and kill expired containers, only do 2 at a time to
+        # reduce lateny if many containers expire
+        tmp_queue = State.reaper_queue.copy()
+        killed = 0
+        ror container_name in tmp_queue:
```
Comment (ajlacey) [inline flasksrc/session.py:228]: for

File: flasksrc/session.py (hunk 1 above)
Comment (blazebissar) [reply flasksrc/session.py]: Added

File: flasksrc/session.py (hunk 4 above)
Comment (blazebissar) [reply flasksrc/session.py:228]: Good catch

**Review by ajlacey — APPROVED** [review]
lgtm

**Pull request merged**